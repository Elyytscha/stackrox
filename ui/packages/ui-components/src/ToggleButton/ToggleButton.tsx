import React, { ReactElement } from 'react';
import { IconProps } from 'react-feather';

export type ToggleButtonProps = {
    icon?: React.FC<IconProps>;
    value: string;
    text: string;
    onClick: (value: string) => void;
    isActive?: boolean;
};

const buttonClassName =
    'first:rounded-l last:rounded-r bg-base-100 border-2 border-base-400 font-600 inline-flex items-center justify-center m-0 py-1 px-2 text-sm uppercase';
const inactiveButtonClassName = `${buttonClassName} first:border-r-0 last:border-l-0 hover:bg-base-200 hover:text-base-700 text-base-600`;
const activeButtonClassName = `${buttonClassName} text-primary-800 bg-primary-300 border-primary-600`;

function ToggleButton({ isActive, icon, value, text, onClick }: ToggleButtonProps): ReactElement {
    const className = isActive ? activeButtonClassName : inactiveButtonClassName;
    const Icon = icon;

    function onClickHandler(event: React.MouseEvent<HTMLButtonElement>): void {
        if (!isActive) {
            const element = event.target as HTMLButtonElement;
            onClick(element.value);
        }
    }

    return (
        <button
            className={className}
            type="button"
            onClick={onClickHandler}
            value={value}
            data-testid={isActive ? 'active-toggle-button' : 'toggle-button'}
        >
            {Icon && <Icon className="h-3 w-3 mr-1" />}
            {text}
        </button>
    );
}

export default ToggleButton;
