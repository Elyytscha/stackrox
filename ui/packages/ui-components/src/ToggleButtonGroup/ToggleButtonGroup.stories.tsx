import React, { useState } from 'react';
import { Meta, Story } from '@storybook/react/types-6-0';
import { Lock, Unlock } from 'react-feather';

import ToggleButtonGroup from './ToggleButtonGroup';
import ToggleButton from '../ToggleButton';

export default {
    title: 'ToggleButtonGroup',
    component: ToggleButtonGroup,
} as Meta;

const VALUE = {
    LOCK: 'LOCK',
    UNLOCK: 'UNLOCK',
};

export const BasicUsage: Story = () => {
    const [activeToggleButton, setActiveToggleButton] = useState(VALUE.LOCK);
    return (
        <ToggleButtonGroup activeToggleButton={activeToggleButton}>
            <ToggleButton value={VALUE.LOCK} text="Lock" onClick={setActiveToggleButton} />
            <ToggleButton value={VALUE.UNLOCK} text="Unlock" onClick={setActiveToggleButton} />
        </ToggleButtonGroup>
    );
};

export const WithIcons: Story = () => {
    const [activeToggleButton, setActiveToggleButton] = useState(VALUE.LOCK);
    return (
        <ToggleButtonGroup activeToggleButton={activeToggleButton}>
            <ToggleButton
                icon={Lock}
                value={VALUE.LOCK}
                text="Lock"
                onClick={setActiveToggleButton}
            />
            <ToggleButton
                icon={Unlock}
                value={VALUE.UNLOCK}
                text="Unlock"
                onClick={setActiveToggleButton}
            />
        </ToggleButtonGroup>
    );
};
