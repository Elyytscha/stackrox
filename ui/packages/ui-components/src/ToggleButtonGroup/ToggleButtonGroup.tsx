import React, { ReactElement } from 'react';

import { ToggleButtonProps } from '../ToggleButton';

type ToggleButtonChild = {
    props: ToggleButtonProps;
};

export type ToggleButtonGroupProps = {
    activeToggleButton: string;
    children: ToggleButtonChild[];
};

// @TODO: See if we can replace the usage of the RadioButtonGroup Component for boolean policy logic with this one
function ToggleButtonGroup({ activeToggleButton, children }: ToggleButtonGroupProps): ReactElement {
    const modifiedChildren = React.Children.map(children, (child: ToggleButtonChild) => {
        if (React.isValidElement(child)) {
            return React.cloneElement(child, {
                isActive: activeToggleButton === child.props.value,
            });
        }
        return child;
    });
    return <div>{modifiedChildren}</div>;
}

export default ToggleButtonGroup;
