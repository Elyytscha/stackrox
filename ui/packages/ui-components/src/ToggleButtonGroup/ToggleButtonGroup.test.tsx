import React, { useState } from 'react';
import { render, fireEvent } from '@testing-library/react';

import ToggleButtonGroup from './ToggleButtonGroup';
import ToggleButton from '../ToggleButton';

function Component() {
    const VALUE = {
        LOCK: 'LOCK',
        UNLOCK: 'UNLOCK',
    };
    const [activeToggleButton, setActiveToggleButton] = useState(VALUE.LOCK);
    return (
        <ToggleButtonGroup activeToggleButton={activeToggleButton}>
            <ToggleButton value={VALUE.LOCK} text="Lock" onClick={setActiveToggleButton} />
            <ToggleButton value={VALUE.UNLOCK} text="Unlock" onClick={setActiveToggleButton} />
        </ToggleButtonGroup>
    );
}

describe('ToggleButtonGroup', () => {
    test('renders the toggle buttons correctly', () => {
        const { container } = render(<Component />);

        expect(container).toMatchSnapshot();
    });

    test('toggles the active button when clicking the inactive button', () => {
        const { getByText, getByTestId } = render(<Component />);

        expect(getByTestId('active-toggle-button')).toHaveTextContent('Lock');

        fireEvent(
            getByText('Unlock'),
            new MouseEvent('click', {
                bubbles: true,
                cancelable: true,
            })
        );

        expect(getByTestId('active-toggle-button')).toHaveTextContent('Unlock');
    });
});
