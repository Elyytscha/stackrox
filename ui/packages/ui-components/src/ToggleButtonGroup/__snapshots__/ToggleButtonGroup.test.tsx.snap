// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleButtonGroup renders the toggle buttons correctly 1`] = `
<div>
  <div>
    <button
      class="first:rounded-l last:rounded-r bg-base-100 border-2 border-base-400 font-600 inline-flex items-center justify-center m-0 py-1 px-2 text-sm uppercase text-primary-800 bg-primary-300 border-primary-600"
      data-testid="active-toggle-button"
      type="button"
      value="LOCK"
    >
      Lock
    </button>
    <button
      class="first:rounded-l last:rounded-r bg-base-100 border-2 border-base-400 font-600 inline-flex items-center justify-center m-0 py-1 px-2 text-sm uppercase first:border-r-0 last:border-l-0 hover:bg-base-200 hover:text-base-700 text-base-600"
      data-testid="toggle-button"
      type="button"
      value="UNLOCK"
    >
      Unlock
    </button>
  </div>
</div>
`;
